### PR TITLE
feat(data-exploration): HogQL filters again

### PR DESCRIPTION
**Pull request by mariusandra**
## Problem

It would be great to use HogQL in property filters.

This is a lighter version of https://github.com/PostHog/posthog/pull/13264/ . Probably makes a lot of old tickets obsolete.

## Changes

Adds support for HogQL filters in a lot of places. Everything is behind the [`hogql_expressions`](https://app.posthog.com/feature_flags/6072) feature flag. 

image

HogQL support in insights, including access to event and person properties:
- [x] trends global filters - works & has test
- [x] trends step filters - works & has test
- [x] trends breakdown - works & has test
- [x] funnels global filters - works & has test
- [x] funnels steps - works & has test
- [x] lifecycle global filters - works & has test
- [x] paths global filters - works & has test
- [x] retention global filters - works & has test
- [x] stickiness - works & has test

HogQL support elsewhere:
- [x] hogql in action step filters, show action on events list - works & has test
- [x] action with hogql in property as trend steps - works, has test
- [x] session recording list - works, has test
- [x] hogql in elements query - works, just not person.properties unless person-on-events is enabled
- [ ] action with hogql in property as part of a cohort - seems to work, but no explicit tests
- [ ] action with hogql in property as funnel step - I'm not sure it works

Technical changes:

- Passes the `hogql_context` object deeply into queries. 
   - This context contains a `values: Dict[str, Any]`, which collects encountered string literals in HogQL
   - Every new entry into `values` is just `hogql_context.values[f"val_{val.length}"] = 'string'`, so the keys are rather predictable.
 - These `hogql_context.values` need to be passed onto the SQL query, and hence get passed around a lot via `hogql_context=filter.hogql_context` or equivalent code. That's where a lot of the changes come in.
- Connect the `HogQLContext` to a `Filter`, to make it easier to pass the context around queries. Note that calling `filter.with_data` (cloning a filter) passes the context by reference, meaning both the old and new filter will gather variables into the same dict.
- Adds support for "global filters" onto the lifecycle chart. It worked in the backend, but was hidden in the frontend 🤔 
- Adds a lot of tests to make sure hogql person and event property access works in all the different insights and other conditions (e.g. as part of an action).
- Removed a dead cohort query from `get_entity_cohort_subquery` that I just couldn't find used anywhere
- Patches a few tests with `@override_settings(PERSON_ON_EVENTS_OVERRIDE=False)  # :KLUDGE: avoid making a bunch of extra queries that may be cached instance-wide`. Forcing this setting avoids all manner of SQL queries (up to 5 per test) that run when `is_cloud()`, `posthoganalytics.feature_enabled()` and `actor_on_events_ready()` (checks async migrations) functions are called. Since we're going to "solve" the person on events issue soon, and I couldn't work out a reliable pattern between how to mock all of this with all the tests that run in CI, I decided to patch 5 test files and tell them to consider person on events disabled. For now.

## How did you test this code?

Work in progress

### CODE REVIEW

**Comment by mariusandra** [comment]
This monster is finally ready for a review. Sorry 😅.

**Review by pauldambra — COMMENTED** [review]
I've never tried to use hogql before and got errors with no feedback from the UI



A bit overwhelming to get the error on every background refresh. We should probably stop refreshing when we know there are errors (not limited to this PR) but hard to know from the UI if I had caused the problem with what I typed , because of my local setup, or because hogql was broken

File: posthog/models/filters/base_filter.py
Comment (mariusandra) [reply posthog/models/filters/base_filter.py:53]: I wrote above "Note that calling filter.with_data (cloning a filter) passes the context by reference, meaning both the old and new filter will gather variables into the same dict."

That's on purpose, as you might compose your query from multiple subqueries, and you'd want to share the variable collector between those and the host query. This happens with e.g. retention.

File: ee/clickhouse/models/test/test_action.py
Comment (mariusandra) [reply ee/clickhouse/models/test/test_action.py:21]: Sure, but it's a helper method in a test that already breaks various contracts 🙃. Normally this would be a util/query function somewhere that does all of this internally.

Comment (pauldambra) [reply ee/clickhouse/models/test/test_action.py:21]: 😅

**Comment by mariusandra** [comment]
Thanks for reviewing @pauldambra !

Re errors: [there's a PR](https://github.com/PostHog/posthog/pull/13703) out to fix this... or would you like me to inline it here? 🤣 I'm also going to work on client side hogql error validation this sprint, which would make things even nicer. We should definitely stop the automatic polling in the events table when the previous result fails... though that might also be outside the scope of "add hogql" 🤔 

In this case though, replace `=` with `==` and the things comparison works. Also if the filter expression returns a string, it's not automatically cast to a boolean for the filter, but assumed to be an error (clickhouse assumes as much). All that will improve... :)

File: posthog/models/filters/base_filter.py
Comment (pauldambra) [reply posthog/models/filters/base_filter.py:53]: 🤯 that feels like a difficult to diagnose bug waiting to happen... coupling on instance is the most evil form of coupling 😰 

At the least the comment in `with_data` needs updating so that intellisense/scanning the code makes it clearer that the instance is shared. It's not making a copy anymore if we have `Filter instance A` <-> `hogql instance` <-> `Filter instance B`

**Review by pauldambra — APPROVED** [review]
> In this case though, replace = with == and the things comparison works

Yep, I think validation on this will be super important and really hard 😅 so wouldn't block this for it. 

-----

I'm fine with progress over perfection here since this is under active development. But the hogql_context instance coupling is scary to me. Since it changes Filters from a bag of data to a thing with a lifecycle...

**Comment by mariusandra** [comment]
> I'm fine with progress over perfection here since this is under active development. But the hogql_context instance coupling is scary to me. Since it changes Filters from a bag of data to a thing with a lifecycle...

I initially didn't connect it to filters at all, but piggybacking off of that saved ~25 changed files :D.

In the case of a retention query (or was it lifecycle?), the filter is cloned for both parts of the queries, and the values are shared. I kept the name of the cloning function as `with_data`, since that's how it was... though renaming that to `shallow_clone` probably clarifies everything.

File: posthog/models/filters/base_filter.py
Comment (mariusandra) [reply posthog/models/filters/base_filter.py:53]: Open to suggestions on how to improve this. I suggested `shallow_clone` below, and [here's a PR](https://github.com/PostHog/posthog/pull/13918) that changes `with_data` to `shallow_clone`. I made a new PR since adding another 41 changed files to this one seems... audacious. 🤣

File: posthog/models/filters/base_filter.py
Comment (mariusandra) [reply posthog/models/filters/base_filter.py:14]: Ideally this HogQL context should be totally ignored in the cache key.

File: ee/clickhouse/queries/test/test_lifecycle.py
Comment (yakkomajuri) [inline ee/clickhouse/queries/test/test_lifecycle.py:212]: Nit: might be more useful to test actually getting some data back?

Comment (yakkomajuri) [inline ee/clickhouse/queries/test/test_lifecycle.py:212]: Nit: might be more useful to test actually getting some data back?

File: posthog/api/test/dashboards/test_dashboard.py
Comment (yakkomajuri) [inline posthog/api/test/dashboards/test_dashboard.py:24]: does this prevent the persons on events test suite from running correctly for these tests?

File: posthog/hogql/hogql.py
Comment (yakkomajuri) [inline posthog/hogql/hogql.py:337]: should we be escaping this here?

File: posthog/hogql/hogql.py
Comment (yakkomajuri) [inline posthog/hogql/hogql.py:333]: let's maybe get even more unique here? would a) help debug b) ensure we don't mess up in the future.

Why not `hogql_val_` as a prefix?

**Review by yakkomajuri — COMMENTED** [review]
Overall looks good although I feel like I'm only really able to review on the surface given a bit of a lack of context.

From the feedback side, appreciate how much testing is here. On the other hand, would have loved this to be broken down into smaller PRs. The files/lines changed number doesn't matter imo, it's more about how there's a list of 7 "technical changes". Makes it easier / more appealing to review if the PR is more contained.

Exciting stuff though!

File: posthog/models/filters/base_filter.py
Comment (pauldambra) [reply posthog/models/filters/base_filter.py:14]: Need to see how that pans out when we're able to add these data tables to e.g. a shared dashboard. 🤔

File: posthog/hogql/hogql.py
Comment (mariusandra) [reply posthog/hogql/hogql.py]: Yeah, let's be explicit with the name, `hogql_val_` works for me.

File: posthog/hogql/hogql.py
Comment (mariusandra) [reply posthog/hogql/hogql.py]: Looking back, this is actually is a remnant of an earlier time, before I added this value collection everywhere. The `format` strings just made it too risky to proceed. I left this in as an option, but there's no real reason to keep it. Removed.

File: ee/clickhouse/queries/test/test_lifecycle.py
Comment (mariusandra) [reply ee/clickhouse/queries/test/test_lifecycle.py:212]: yup, makes sense

File: posthog/api/test/dashboards/test_dashboard.py
Comment (mariusandra) [reply posthog/api/test/dashboards/test_dashboard.py:24]: ugh, yes, I think it does. `settings.PERSON_ON_EVENTS_OVERRIDE` is checked before `PERSON_ON_EVENTS_ENABLED` env that's enabled in the `.github` folder. I'll think about it, but this is obviously a no go as is.

File: posthog/api/test/test_insight.py
Comment (macobo) [inline posthog/api/test/test_insight.py:2375]: What is this comment about?

File: posthog/api/test/test_insight.py
Comment (macobo) [inline posthog/api/test/test_insight.py:2359]: Could these be separate tests as the comment indicates?

File: posthog/api/test/test_event.py
Comment (macobo) [inline posthog/api/test/test_event.py:65]: Why is license being fetched 4x in one request now?

File: posthog/api/test/test_insight.py
Comment (mariusandra) [reply posthog/api/test/test_insight.py:2375]: seems to be a copy/pasta. Will yeet.

File: posthog/api/test/dashboards/test_dashboard.py
Comment (macobo) [inline posthog/api/test/dashboards/test_dashboard.py:151]: I'm concerned about this given I recently worked on it (though I need to resurrect the PR) - what is the 'false' N+1 being added here and is it truly false?

File: ee/clickhouse/queries/test/test_cohort_query.py
Comment (macobo) [inline ee/clickhouse/queries/test/test_cohort_query.py:2219]: This is a smell - params handling is already a major painpoint and I'd be careful about making it worse. Why doesn't CohortQuery do this itself?

File: posthog/api/test/test_event.py
Comment (mariusandra) [reply posthog/api/test/test_event.py:65]: That's what `team.person_on_events_querying_enabled` does. Every time we make a query with hogql "properties.bla" in it, we make a request to check async migration, the license, etc. This should all be cached in production, but there's an explicit `not TEST` in some of the caching code.

File: posthog/queries/retention/actors_query.py
Comment (macobo) [inline posthog/queries/retention/actors_query.py:66]: Why isn't this done by actor query itself?

Problem is - writing code you now need to know that _sometimes_ you need to inject some params in from somewhere but other times you do not.

File: posthog/queries/funnels/base.py
Comment (macobo) [inline posthog/queries/funnels/base.py:754]: Should this function now be returning params as well?

File: posthog/models/filters/filter.py
Comment (macobo) [inline posthog/models/filters/filter.py:128]: Would it make sense move this to a mixin and make it a @cached_property instead?

File: posthog/api/test/dashboards/test_dashboard.py
Comment (mariusandra) [reply posthog/api/test/dashboards/test_dashboard.py:151]: Each query we make adds another insight that then gets `team.person_on_events_querying_enabled` called when it eventually gets queried as part of the dashboard.

I literally spent hours figuring out how to not get these extra queries for persons on events. This approach I settled on (exclude the N+1 tasks, add query counts elsewhere) is the least evil of all I tried. Snapshots won't work well either because each license test checks `valid_until > $inlined_timestamp`...

File: posthog/queries/retention/actors_query.py
Comment (mariusandra) [reply posthog/queries/retention/actors_query.py:66]: It was hard to understand how all these queries work --> which of these get inlined as subqueries, which are full queries. This means I was never sure what's the *right time* to inline the hogql values into the params. Hence I took an alternative approach: the entire HogQL value gathering is something that is expected to be controlled outside the queries. I don't mind re-evaluating this approach, as the worse case for that is just passing a lot of duplicate (yet still unique) params around.

File: posthog/queries/funnels/base.py
Comment (mariusandra) [reply posthog/queries/funnels/base.py:754]: Not "now", as it gathers the params into `hogql_context.values`, which is shared/controlled outside the queries. At least that's how it is "now".

File: posthog/models/filters/filter.py
Comment (mariusandra) [reply posthog/models/filters/filter.py]: Sure, let's try.

File: posthog/queries/retention/actors_query.py
Comment (macobo) [reply posthog/queries/retention/actors_query.py:66]: Fair - I can get why we got here. We'll need to revisit this soonish and push these params 'deeper' again (it's a maintanence nightmare otherwise) but not sure we should block this PR further.

Comment (mariusandra) [reply posthog/queries/retention/actors_query.py:66]: I'm happy to take on cleaning up all these queries in a followup... or even here if you think it's blocking, though I'd obviously prefer to do a followup. For now, the worst thing that can happen in a future query that forgets to add these values... is a missing `%(hogql_val_1)s` error. 

It seems very possible to find a clear boundary between queries and subqueries, and inline those values into queries there. Those big query classes like `Retention`, `ClickhouseFunnelTrends`, etc might be just that boundary... but then it's not predefined where their `Filter`-s come from... blergh... 🤔

**Review by macobo — APPROVED** [review]
Let's go (once green)!

The params stuff is IMO the weakest part of this - we're breaking a long-held convention - but keeping this PR outstanding is doing more harm right now. Let's rely on tests to refactor soon.

**Comment by mariusandra** [comment]
Thanks! I'm very happy to keep working on this after the PR lands. However, I'm now running into this in one of the CI tests:

```
/home/runner/work/_temp/f160cbea-1f86-4285-a8e1-4d085f6d9f94.sh: line 4:  5600 Aborted                 (core dumped) pytest ee -m "not async_migrations" --splits 5 --group 2 --durations=100 --durations-min=1.0 --store-durations $PYTEST_ARGS
```

🤔 

Perhaps busting the cache... ?

**Comment by mariusandra** [comment]
image

🤣

File: ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel.ambr
Comment (mariusandra) [inline ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel.ambr:6]: This is a concerning change 🤔

**Pull request merged**
